File: test/vectors/cross-emitter-matrix/Verify.java
// Cross-emitter matrix — Java verifier.
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.security.*;
import java.security.spec.X509EncodedKeySpec;
import java.util.*;
import com.fasterxml.jackson.databind.*;
import com.fasterxml.jackson.databind.node.*;

public class Verify {
    static final String[] FIELD_ORDER = {
        "id","timestamp","method","toolName","namespace","upstream",
        "principal","durationMs","success","errorCode","previousHash"
    };
    static final ObjectMapper M = new ObjectMapper();

    static Object canonicalizeValue(Object v) {
        if (v == null) return NullNode.getInstance();
        if (v instanceof NullNode) return NullNode.getInstance();
        if (v instanceof BooleanNode) return v;
        if (v instanceof TextNode) return v;
        if (v instanceof IntNode || v instanceof LongNode) {
            long i = ((JsonNode)v).longValue();
            if (Math.abs(i) > (1L << 53) - 1) throw new RuntimeException("unsafe number " + i);
            return v;
        }
        if (v instanceof DoubleNode || v instanceof FloatNode) {
            throw new RuntimeException("unsafe number (float): " + v);
        }
        if (v instanceof ArrayNode) {
            ArrayNode inner = M.createArrayNode();
            for (JsonNode x : (ArrayNode)v) inner.add((JsonNode)canonicalizeValue(x));
            ArrayNode tagged = M.createArrayNode();
            tagged.add("L");
            tagged.add(inner);
            return tagged;
        }
        if (v instanceof ObjectNode) {
            ObjectNode obj = (ObjectNode)v;
            List<String> keys = new ArrayList<>();
            obj.fieldNames().forEachRemaining(keys::add);
            keys.sort((a,b) -> {
                byte[] ba = a.getBytes(StandardCharsets.UTF_16BE);
                byte[] bb = b.getBytes(StandardCharsets.UTF_16BE);
                int len = Math.min(ba.length, bb.length);
                for (int i = 0; i < len; i++) {
                    int cmp = (ba[i] & 0xff) - (bb[i] & 0xff);
                    if (cmp != 0) return cmp;
                }
                return ba.length - bb.length;
            });
            ArrayNode pairs = M.createArrayNode();
            for (String k : keys) {
                ArrayNode pair = M.createArrayNode();
                pair.add(k);
                pair.add((JsonNode)canonicalizeValue(obj.get(k)));
                pairs.add(pair);
            }
            ArrayNode tagged = M.createArrayNode();
            tagged.add("M");
            tagged.add(pairs);
            return tagged;
        }
        throw new RuntimeException("unsupported type " + v.getClass());
    }

    static String canonicalize(JsonNode record) throws Exception {
        ArrayNode ordered = M.createArrayNode();
        for (String k : FIELD_ORDER) {
            JsonNode v = record.get(k);
            ArrayNode pair = M.createArrayNode();
            pair.add(k);
            pair.add((JsonNode)canonicalizeValue(v == null ? NullNode.getInstance() : v));
            ordered.add(pair);
        }
        for (String opt : new String[]{"decisionContextDigest", "extensionsDigest", "aiInvocation", "parties"}) {
            JsonNode v = record.get(opt);
            if (v != null && !v.isNull()) {
                ArrayNode pair = M.createArrayNode();
                pair.add(opt);
                pair.add((JsonNode)canonicalizeValue(v));
                ordered.add(pair);
            }
        }
        return M.writeValueAsString(ordered);
    }

    static final byte[] SPKI_PREFIX = new byte[]{
        0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x03, 0x21, 0x00
    };

    static ObjectNode verifyOne(JsonNode payload, KeyFactory kf) throws Exception {
        String pubHex = payload.get("public_key_hex").asText();
        String sigHex = payload.get("signature_hex").asText();
        JsonNode record = payload.get("record");

        // Build SPKI DER: 12-byte header + 32-byte pubkey
        byte[] spki = new byte[12 + 32];
        System.arraycopy(SPKI_PREFIX, 0, spki, 0, 12);
        for (int i = 0; i < 32; i++) spki[12 + i] = (byte)Integer.parseInt(pubHex.substring(i*2, i*2+2), 16);

        X509EncodedKeySpec keySpec = new X509EncodedKeySpec(spki);
        PublicKey pubKey = kf.generatePublic(keySpec);

        byte[] sigBytes = new byte[64];
        for (int i = 0; i < 64; i++) sigBytes[i] = (byte)Integer.parseInt(sigHex.substring(i*2, i*2+2), 16);

        boolean verified = false;
        String canonical = null;
        try {
            canonical = canonicalize(record);
            Signature sig = Signature.getInstance("Ed25519");
            sig.initVerify(pubKey);
            sig.update(canonical.getBytes(StandardCharsets.UTF_8));
            verified = sig.verify(sigBytes);
        } catch (Exception e) {
            canonical = "ERROR: " + e.getMessage();
        }
        ObjectNode out = M.createObjectNode();
        out.put("verified", verified);
        out.put("local_canonical", canonical);
        out.put("sig_hex", sigHex);
        return out;
    }

    public static void main(String[] args) throws Exception {
        // Warm KeyFactory reused across records in daemon mode.
        KeyFactory kf = KeyFactory.getInstance("Ed25519");

        String daemon = System.getenv("DAEMON_MODE");
        if (daemon != null && daemon.equals("1")) {
            // Daemon mode: id-wrapped envelopes {"id": "...", "record": {...},
            // "signature_hex": "...", "public_key_hex": "..."} on stdin, id-wrapped
            // JSON responses on stdout. Loop until EOF; a per-record failure emits
            // {"id": <id>, "ok": false, "error": "..."} and continues.
            BufferedReader in = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
            PrintWriter out = new PrintWriter(new OutputStreamWriter(System.out, StandardCharsets.UTF_8));
            String line;
            while ((line = in.readLine()) != null) {
                if (line.isEmpty()) continue;
                String id = null;
                try {
                    JsonNode envelope = M.readTree(line);
                    JsonNode idNode = envelope.get("id");
                    id = (idNode == null || idNode.isNull()) ? null : idNode.asText();
                    // verifyOne reads record/signature_hex/public_key_hex from the payload;
                    // the extra "id" key in the envelope is ignored by it.
                    ObjectNode verified = verifyOne(envelope, kf);
                    ObjectNode result = M.createObjectNode();
                    if (id != null) result.put("id", id); else result.putNull("id");
                    result.put("ok", true);
                    result.put("verified", verified.get("verified").asBoolean());
                    result.set("local_canonical", verified.get("local_canonical"));
                    out.println(M.writeValueAsString(result));
                } catch (Exception e) {
                    ObjectNode err = M.createObjectNode();
                    if (id != null) err.put("id", id); else err.putNull("id");
                    err.put("ok", false);
                    err.put("error", e.getMessage() == null ? e.getClass().getSimpleName() : e.getMessage());
                    out.println(M.writeValueAsString(err));
                }
                out.flush();
            }
            return;
        }

        // One-shot mode: single payload on stdin, single JSON object on stdout (no newline).
        JsonNode payload = M.readTree(System.in);
        ObjectNode result = verifyOne(payload, kf);
        System.out.print(M.writeValueAsString(result));
    }
}

File: test/vectors/cross-emitter-matrix/verify.swift
// Cross-emitter matrix — Swift verifier.
import Foundation
import CryptoKit

let FIELD_ORDER = [
  "id","timestamp","method","toolName","namespace","upstream",
  "principal","durationMs","success","errorCode","previousHash"
]

func hex2bytes(_ s: String) -> Data {
    var d = Data(capacity: s.count / 2)
    var i = s.startIndex
    while i < s.endIndex {
        let end = s.index(i, offsetBy: 2)
        d.append(UInt8(s[i..<end], radix: 16)!)
        i = end
    }
    return d
}

func compactJSON(_ v: Any) -> String {
    let d = try! JSONSerialization.data(withJSONObject: v, options: [.withoutEscapingSlashes])
    return String(data: d, encoding: .utf8)!
}

func sortKeysUTF16(_ keys: [String]) -> [String] {
    return keys.sorted { a, b in
        let ua = Array(a.utf16)
        let ub = Array(b.utf16)
        let len = min(ua.count, ub.count)
        for k in 0..<len {
            if ua[k] < ub[k] { return true }
            if ua[k] > ub[k] { return false }
        }
        return ua.count < ub.count
    }
}

func canonicalizeValue(_ v: Any?) -> Any {
    if v == nil { return NSNull() }
    if let n = v as? NSNull { return n }
    if let s = v as? String { return s }
    if let n = v as? NSNumber {
        if CFGetTypeID(n) == CFBooleanGetTypeID() { return n.boolValue }
        let objType = String(cString: n.objCType)
        if objType == "d" || objType == "f" { fatalError("unsafe float \(n)") }
        let i = n.int64Value
        precondition(abs(i) <= (1 << 53) - 1, "unsafe integer \(i)")
        return i
    }
    if let a = v as? [Any] {
        return ["L", a.map { canonicalizeValue($0) }]
    }
    if let m = v as? [String: Any] {
        let keys = sortKeysUTF16(Array(m.keys))
        let pairs: [Any] = keys.map { k in [k, canonicalizeValue(m[k] as Any)] }
        return ["M", pairs]
    }
    fatalError("unsupported type \(type(of: v!))")
}

func canonicalize(_ record: [String: Any]) -> String {
    var ordered: [[Any]] = FIELD_ORDER.map { k -> [Any] in
        let v = record[k] ?? NSNull()
        return [k, canonicalizeValue(v)]
    }
    for opt in ["decisionContextDigest", "extensionsDigest", "aiInvocation", "parties"] {
        if let v = record[opt], !(v is NSNull) {
            ordered.append([opt, canonicalizeValue(v)])
        }
    }
    return compactJSON(ordered)
}

let inputData = FileHandle.standardInput.readDataToEndOfFile()
let payload = try JSONSerialization.jsonObject(with: inputData) as! [String: Any]
let record = payload["record"] as! [String: Any]
let sigHex = payload["signature_hex"] as! String
let pubHex = payload["public_key_hex"] as! String

let pubKey = try Curve25519.Signing.PublicKey(rawRepresentation: hex2bytes(pubHex))
let sig = hex2bytes(sigHex)

var verified = false
var canonical: String? = nil
do {
    canonical = canonicalize(record)
    verified = pubKey.isValidSignature(sig, for: canonical!.data(using: .utf8)!)
} catch {
    canonical = "ERROR: \(error)"
}

let out: [String: Any] = [
    "verified": verified,
    "local_canonical": canonical ?? "null",
    "sig_hex": sigHex,
]
let outData = try JSONSerialization.data(withJSONObject: out)
FileHandle.standardOutput.write(outData)
